### PR TITLE
REF: implement Categorical._box_func, make _box_func a method

### PR DESCRIPTION
**Pull request by jbrockmendel**
Defining `_box_func` on `Categorical` is to facilitate more code sharing with datetimelike.  Also there are a bunch of places that do something _similar_ but slightly different that I want to standardize within `Categorical`.

Changing `_box_func` from a property-lambda to a method on the datetimelike arrays gives a slight perf boost:

```
In [1]: import pandas as pd                                                     
In [2]: dti = pd.date_range("2016-01-01", periods=3)                 
           
In [3]: %timeit dti[1]                                                          
8.73 µs ± 50.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- master
8.08 µs ± 119 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)   # <-- PR

```

### CODE REVIEW

File: pandas/core/arrays/timedeltas.py [hunk 1]
```diff
@@ -108,9 +108,8 @@ class TimedeltaArray(dtl.DatetimeLikeArrayMixin, dtl.TimelikeOps):
     # Note: ndim must be defined to ensure NaT.__richcmp(TimedeltaArray)
     #  operates pointwise.
 
-    @property
-    def _box_func(self):
-        return lambda x: Timedelta(x, unit="ns")
+    def _box_func(self, x):
```
Comment (jreback) [inline pandas/core/arrays/timedeltas.py:111]: ideally can you type the returns

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, ok if you want to do as a followon as well

**Comment by jbrockmendel** [comment]
annotated+green

**Review by jreback — APPROVED** [review]

File: pandas/core/arrays/period.py [hunk 2]
```diff
@@ -484,9 +484,8 @@ def _time_shift(self, periods, freq=None):
             values[self._isnan] = iNaT
         return type(self)(values, freq=self.freq)
 
-    @property
-    def _box_func(self):
-        return lambda x: Period._from_ordinal(ordinal=x, freq=self.freq)
+    def _box_func(self, x) -> Union[Period, NaTType]:
```
Comment (jreback) [inline pandas/core/arrays/period.py:487]: should alias this kind of type at somepoint, maybe PeriodOrNaT (and the others as well)

**Review by jreback — APPROVED** [review]

**Pull request merged**